src/NumberTheory/vinberg.jl

```
@@ -353,3 +353,156 @@ function vinberg_algorithm(S::ZZLat, upper_bound; v0=QQ[0;]::QQMatrix, root_leng
   return [r * B for r in roots]
 end
 
+
+function _get_tau(f::QQMatrix, Qb)
+  tau = zero(Qb)
+  for lambda in eigenvalues(Qb, f)
+    if abs(lambda)>tau
+      tau = abs(lambda)
+    end
+  end
+  return tau
+end
+
+function _get_bilinearform(L::ZZLat, Qb)
+  gram_in_Qb = change_base_ring(Qb, gram_matrix(ambient_space(L)))
+  #return (a, b)->inner_product(ambient_space(L), a,b)[1,1] -> it doesn't go well as there is no convienient way to convert ZZLat in QQ field to the Qb field
+  return (a,b)-> ((change_base_ring(Qb, a))*gram_in_Qb*transpose(change_base_ring(Qb, b)))[1]
+end
+
+function _get_C0(Lf::ZZLatWithIsom, tau::QQBarFieldElem)
+  charPolyF = characteristic_polynomial(Lf)
+  x = gen(parent(charPolyF))
+  (n, remainder) = remove(charPolyF, x-1) # remove all x-1 factors
+  return div(remainder, minpoly(parent(charPolyF), tau)) #remove Salem polynomial of salem number tau
+end
+
+function _get_Cfancy(Lf::ZZLatWithIsom, C0)
+  return short_vectors_iterator(lattice(kernel_lattice(Lf, C0)), 2 , 2)
+end
+
+
+@doc raw"""
+    _get_h(L::ZZLat, v, w, Qb, bi_form) -> QQMatrix
+
+Return a vector $h$ in the lattice `L`, such that $bi_form(h,h)>0$. 
+
+Vectors 'v','w' are eigenvectors of isometry assotiated with L based on salem number $tau$ and $tau^(-1)$
+
+'Qb' is a current field, where all calculations happens
+"""
+function _get_h(L::ZZLat, v, w, Qb, bi_form)
+  RR = ArbField(64)
+  l = number_of_rows(basis_matrix(L))
+  h0 = (v+w)*change_base_ring(Qb, inv(basis_matrix(L)))
+  n = 1
+  i = 1
+  z = matrix(Qb,1,l,rand(-10:10, l)) # random small vector in lattice basis
+  h = (z+h0*n) # in lattice basis
+  h = map(x->QQ(ZZ(floor(RR(x)))) , h)*basis_matrix(L) # in ambient space basis and rounded
+  while bi_form(h,h) <= 0 || n == 10000  # block on 10000, because for a bigger n time consumption is way too big
```

please throw an error if the while loop fails

---

Added